### ISSUE TITLE
lint: warn about identical conditions inside and outside of functions

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang-nursery/rust-clippy#2278]
There could be a warning if something is checked outside of a function, passed inside the function and then checked inside the function again.
This might spot redundant conditions but also logic errors if the conditions are conflicting.

````rust


fn fun(awesomestruct: SomeStruct) {
   if awesomestruct.bar { return } // check inside function
   [...]
}

fn main() {
    let struct: SomeStruct = something(); 

    if struct.bar { // same condition that is inside of function
         fun(struct);
    }

}
````

**Comment by killercup** [comment]
Good idea!

I'd personally also add some help text to suggest moving this check into the type if possible/sensible (like [this](https://play.rust-lang.org/?gist=61f56f2dabc495486e3659553a5707a8&version=stable)) but I'm not sure there is community consensus or good docs on this.

**Comment by matthiaskrgr** [comment]
I would say it depends.
If I have a caller with several consecutive function calls returning early because of the same condition it makes more sense to hoist the check into the caller. The compiler would probably do the same anyway. It makes clear the entire guarded block may not be run at first glance.
